### ISSUE TITLE
Options "--add-subtitle" and "--main-audio" does not seem to work.

### ISSUE THREAD

**Issue by muhrizali** [lisamelton/video_transcoding#350]
Hello, @lisamelton
I have been using your tools for long time now (really love them by the way!). I have encoded thousands of mainly Japanese anime video ".mkv" files and have never encountered this issue. Yesterday, I got a new TV show and decided to use these tools on them to add to my collection.

I was using the following command:
`transcode-video --no-auto-burn --add-subtitle all --main-audio jpn ./input/* -o output/`

All the video files in the input folder are consistent, meaning they all have similar subtitle and audio tracks.

But the output transcoded files does not seem to follow the instructions. There are no subtitles added in the output but JPN audio track was added. 

But later, when I tested the audio track, changing it to `--main-audio eng`, that does not seem to work too, as there was no ENG track in the output file, JPN audio track added again.

I have even scanned the first file from input folder:

```
+ title 1:
  + stream: input/Frieren Beyond Journey's End - S01E01 (WEB 1080p x265 10-bit AAC Opus) [0F7F64F6].mkv
  + duration: 00:26:00
  + size: 1920x1080, pixel aspect: 1/1, display aspect: 1.78, 23.976 fps
  + chapters:
    + 1: duration 00:01:30
    + 2: duration 00:17:27
    + 3: duration 00:05:23
    + 4: duration 00:01:30
    + 5: duration 00:00:11
  + audio tracks:
    + 1,  (Opus, 2.0 ch) (iso639-2: jpn)
    + 2, English (AAC LC, 2.0 ch) (iso639-2: eng)
  + subtitle tracks:
    + 1, English (SSA)
    + 2, English (SSA)
```
I am sure I am missing something here, but I am not sure what. I really appreciate if you could guide me a little. I know that the tool is not meant to be used for multiple files, but I have used the batch-transcode-video script at https://github.com/nwronski/batch-transcode-video from @nwronski and it is giving me the same results.

**Comment by lisamelton** [comment]
@muhrizali Do the `--no-auto-burn --add-subtitle all --main-audio jpn` options work when you apply them to a single input?

**Comment by muhrizali** [comment]
@lisamelton No, the options does not seem to have any effect even in the case of single inputs.

In case it would be helpful, I have found the section in the log file that I think represents the "job configuration" for the encoding task: There are no subtitle tracks added.
```
[18:48:28] Starting Task: Encoding Pass
[18:48:28] work: skipping crop/scale filter
[18:48:28] job configuration:
[18:48:28]  * source
[18:48:28]    + ./input/Frieren Beyond Journey's End - S01E01 (WEB 1080p x265 10-bit AAC Opus) [0F7F64F6].mkv
[18:48:28]    + title 1, chapter(s) 1 to 5
[18:48:28]    + container: matroska,webm
[18:48:28]    + data rate: 17972 kbps
[18:48:28]  * destination
[18:48:28]    + /mnt/storage/Downloads/Frieren Beyond Journey's End - Season 1 (WEB 1080p x265 10-bit AAC Opus) [Dual Audio]/output/Frieren Beyond Journey's End - S01E01 (WEB 1080p x265 10-bit AAC Opus) [0F7F64F6].mkv
[18:48:28]    + container: Matroska (libavformat)
[18:48:28]      + chapter markers
[18:48:28]  * video track
[18:48:28]    + decoder: hevc 10-bit (yuv420p10le)
[18:48:28]    + filters
[18:48:28]      + Framerate Shaper (mode=0)
[18:48:28]        + frame rate: same as source (around 23.976 fps)
[18:48:28]      + Format (format=yuv420p)
[18:48:28]    + Output geometry
[18:48:28]      + storage dimensions: 1920 x 1080
[18:48:28]      + pixel aspect ratio: 1 : 1
[18:48:28]      + display dimensions: 1920 x 1080
[18:48:28]    + encoder: H.264 (libx264)
[18:48:28]      + preset:  medium
[18:48:28]      + options: vbv-maxrate=6000:vbv-bufsize=12000:crf-max=25:qpmax=34
[18:48:28]      + profile: high
[18:48:28]      + level:   4.0
[18:48:28]      + quality: 1.00 (RF)
[18:48:28]      + color profile: 1-1-1
[18:48:28]      + chroma location: left
[18:48:28]  * audio track 1
[18:48:28]    + name: Japanese 2.0 Opus
[18:48:28]    + decoder: 日本語 (Opus, 2.0 ch) (track 1, id 0x1)
[18:48:28]      + samplerate: 48000 Hz
[18:48:28]    + mixdown: Stereo
[18:48:28]    + encoder: AAC (libavcodec)
[18:48:28]      + bitrate: 128 kbps, samplerate: 48000 Hz
Encoding: task 1 of 1, 0.00 %[18:48:28] sync: expecting 37424 video frames
Encoding: task 1 of 1, 0.00 %[18:48:28] encx264: min-keyint: 24, keyint: 240
```

**Comment by muhrizali** [comment]
Just now, I have tried the same command with a different video file wondering whether the issue was in the original ".mkv" file but the outputs in the log file were identical. The commands have no effect there too.

**Comment by muhrizali** [comment]
Man, you are certainly taking your sweet time.

**Comment by muhrizali** [comment]
I have done my own little research and found out that the problem is in the version of HandBrakeCLI. The problem arises with the release of version 1.7.x. 

Though I couldn't be able to test myself, I have read somewhere that previous version series 1.6.x was working fine.

**Comment by muhrizali** [comment]
> I downgraded to HandbrakeCLI 1.6.1 and confirmed that `--add-audio` and `--add-subtitle` flags worked as expected.

I wanted to downgraded to version 1.6.X on my ubuntu Linux but it is a real PITA to build it from the source. Tried it many times with no success.

**Comment by dgaidula** [comment]
Had same issue this am, with longtime bash script that processes files with audio options for commentary, etc. Thought I was losing my mind, but then found this issue and verified that rolling back to HandbrakeCLI 1.6.1 solves it, main-audio flag now works again, etc.

**Comment by lisamelton** [comment]
@muhrizali @dirtystylus @dgaidula I've just released a complete redesign and rewrite of everything in the project so that `transcode-video.rb` _will_ work correctly with the latest version of `HandBrakeCLI`. Sorry that took so long. But can you re-test now with this new version? Thanks.

**Comment by dgaidula** [comment]
@lisamelton I can confirm for my use case—adding names and/or options for special cases via a 'side car' txt file—that my batch script reads. This version is now working for me, although I did have to rework my batch script in a couple place since this new one is not one-to-one drop-in replacement. (I will also have reformat my side-car files, too.)

Just to save others time (as it took me a bit to figure all this out). You have to add audio via '--add-audio', but then use the extra options syntax to name it, etc.

`--add-audio 2 -x aname=,"Commentary: Richard Schickle"`

also note the comma in 'aname' option to skip the naming of the first track since your script already names it 'Surround'. I may go back and name these 'Surround: Main Audio' now that I think of it.

Mildly, unrelated, but why did you abandon your custom rate control for two-pass? 

Also any chance to bring back ac3-640k as an extra option when down mixing? (I know I can probably just update my batch script to override it etc, but it looks like it would be an easy flag to add.)

Last question: am I right to use the 'ca_aac' option if I am transcoding via macOS? (I have a couple minis that are my transcoders.)

As always, much appreciation for all your work in this space and of course, Safari...

**Comment by lisamelton** [comment]
@dgaidula I'm glad it's working for you and that you have embraced the power and flexibility of the `--extra` (or `-x) option.

I replaced my old single-pass ratecontrol system mostly for three reasons. First, over the years I found too many cases of it causing artifacts that I could not mitigate. It didn't always mitigate color banding either. Second, everyone's computer seems to be _so_ much more powerful these days that even two-pass transcoding is faster than real time. It's actually 4x faster on my PC. Third, I can reduce the output bitrate target significantly by using two passes. Try `--bitrate 1` (`-b 1`) when using the default video mode to see how much more you can reduce the target and how good it still looks. That wouldn't be possible with the old ratecontrol system.

As for 640 Kbps (E-)AC-3 audio? No, that won't be an option except through `--extra`. It goes against the idea of compression and the additional bitrate doesn't add significant perceptible quality. My advice is to avoid using Dolby Digital format for your audio anyway since it's so much larger than AAC and doesn't sound noticeably better. The only reason I included Dolby Digital support is for compatibility.

And, yes, you can use `--aac-encoder ca_aac` if you're transcoding on macOS. It's possible that it could improve quality in some cases. However, fair warning, it's likely not to be a noticeable improvement.

**Comment by dirtystylus** [comment]
@lisamelton Took me a while to get back to this but I can confirm that this reworked version is running well against handbrakecli v1.7.3. Thanks so much Lisa for your continued work on these tools 💜

I have opened a separate issue with a question about output file renaming and passing those arguments to handbrakecli via the `--extra` flag: https://github.com/lisamelton/video_transcoding/issues/356

**Comment by lisamelton** [comment]
@dirtystylus Thanks for confirming! And you are very welcome. 💖

I'll close this now.
